pkg/servicedisc/autoconnect.go

```diff
@@ -42,83 +41,83 @@ func MakeConnector(conf Config, maxConns int, tm *transport.Manager, log *loggin
 	connector.log = log
-	connector.conns = make(map[cipher.PubKey]struct{})
 	connector.tm = tm
 	return connector
 }
 
 // Run implements Autoconnector interface
 func (a *autoconnector) Run(ctx context.Context) error {
-	retrier := netutil.NewRetrier(fetchServicesDelay, 0, 2)
 	for {
 		time.Sleep(PublicServiceDelay)
-		a.checkConns()
-		if len(a.conns) == a.maxConns {
-			continue
-		}
-		var services []Service
-		fetch := func() (err error) {
-			// "return" services up from the closure
-			services, err = a.client.Services(ctx, a.maxConns)
-			if err != nil {
-				return err
-			}
-			return nil
-		}
-		if err := retrier.Do(fetch); err != nil {
-			a.log.Errorf("Cannot fetch services: %s", err)
-			return err
+		a.log.Infof("Fetching public visors")
+		addresses, err := a.fetchPubAddresses(ctx)
+		if err != nil {
+			a.log.Errorf("Cannot fetch public services: %s", err)
 		}
 
-		for _, service := range services {
-			pk := service.Addr.PubKey()
-			if _, ok := a.conns[pk]; ok {
+		trs := a.updateTransports()
+		absent := a.calculateAbsent(addresses, trs)
+		for _, pk := range absent {
+			a.log.WithField("pk", pk).Infoln("Adding transport to public visor")
+			logger := a.log.WithField("pk", pk).WithField("type", tptypes.STCPR)
+			if _, err := a.tm.SaveTransport(ctx, pk, tptypes.STCPR, transport.LabelAutomatic); err != nil {
+				logger.WithError(err).Warnln("Failed to add transport to public visor")
 				continue
 			}
-			a.connect(ctx, pk)
+			logger.Infoln("Added transport to public visor")
 		}
 	}
 }
 
-// check if existing connections are still active using checker
-// and delete those that are not
-func (a *autoconnector) checkConns() {
-	toDelete := make([]cipher.PubKey, 0)
-	for pk := range a.conns {
-		if !a.checkConn(pk) {
-			toDelete = append(toDelete, pk)
+// Remove all inactive automatic transports and return all active
+// automatic transports
+func (a *autoconnector) updateTransports() []*transport.ManagedTransport {
+	trs := a.tm.GetTransportsByLabel(transport.LabelAutomatic)
+	var trsActive []*transport.ManagedTransport
+	for _, tr := range trs {
+		if !tr.IsUp() {
+			a.tm.DeleteTransport(tr.Entry.ID)
+		} else {
+			trsActive = append(trsActive, tr)
 		}
 	}
-	for _, pk := range toDelete {
-		delete(a.conns, pk)
-	}
+	return trsActive
 }
 
-func (a *autoconnector) checkConn(pk cipher.PubKey) bool {
-	t, err := a.tm.GetTransport(pk, tptypes.STCPR)
-	if err != nil {
-		return false
+func (a *autoconnector) fetchPubAddresses(ctx context.Context) ([]cipher.PubKey, error) {
+	retrier := netutil.NewRetrier(fetchServicesDelay, 0, 2)
+	var services []Service
+	fetch := func() (err error) {
+		// "return" services up from the closure
+		services, err = a.client.Services(ctx, a.maxConns)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	if err := retrier.Do(fetch); err != nil {
+		return nil, err
 	}
-	up := t.IsUp()
-	if !up {
-		a.tm.DeleteTransport(t.Entry.ID)
+	var pkeys []cipher.PubKey
```

Comment:
Can you check if we use this name elsewhere? I think `pks` or `keys` is more canonical.